src/main/java/soot/dexpler/DexFillArrayDataTransformer.java

```diff
@@ -135,4 +113,50 @@ protected void internalTransform(final Body body, String phaseName, Map<String,
     }
   }
 
+  /**
+   * Check the all available definitions of the current array to detect the array type and thus the type of the data loaded
+   * by the array-fill-data instruction.
+   * 
+   * @param l
+   *          local the array we are interested in is saved in
+   * @param u
+   *          unit we start our search
+   * @param defs
+   * @param arrayTypes
+   *          result list containing the discovered array type(s)
+   */
+  private void checkArrayDefinitions(Local l, Unit u, LocalDefs defs, List<Type> arrayTypes) {
+    List<Unit> assDefs = defs.getDefsOfAt(l, u);
+    for (Unit d : assDefs) {
+      if (d instanceof AssignStmt) {
+        AssignStmt arrayAssign = (AssignStmt) d;
+        Value source = arrayAssign.getRightOp();
+        if (source instanceof NewArrayExpr) {
+          // array is assigned from a newly created array
+          NewArrayExpr newArray = (NewArrayExpr) source;
+          arrayTypes.add(newArray.getBaseType());
+          continue;
+        }
+        if (source instanceof InvokeExpr) {
+          // array is assigned from the return value of a function
+          InvokeExpr invExpr = (InvokeExpr) source;
+          Type aType = invExpr.getMethodRef().getReturnType();
+          if (!(aType instanceof ArrayType)) {
+            throw new InternalError("Failed to identify the array type. The identified method invocation "
+                + "does not return an array type. Invocation: " + invExpr.getMethodRef());
+          }
+          arrayTypes.add(((ArrayType) aType).getArrayElementType());
+          continue;
+        }
+        if (source instanceof Local) {
+          // our array is defined by an assignment from another array => check the definition of that other array.
+          Local newLocal = (Local) source; // local of the "other array"
+          checkArrayDefinitions(newLocal, d, defs, arrayTypes);
```

Comment:
We need a set of assignments that we have already seen to avoid infinite recursion. I am thinking about code like this:

```
if (...) {
a = b;
} else {
b = a;
}
```

Reply:
recursion depth parameter introduced that decreases with every recursion and abort check added.